### DISCUSSION TITLE
[Q,FR] How do I get the documentation for an arbitrary symbol (inputted as a string)?

### DISCUSSION THREAD

**Discussion by NightMachinery** [joaotavora/eglot#744] [Q&A]
I want to implement some UI elements that show documentation. 
E.g., I like to add a docstring to `ivy`. An example follows for elisp:
![image](https://user-images.githubusercontent.com/36224762/134042304-e2755214-6c69-4396-9700-a05233d6aaf0.png)

Here is an eglot-managed Julia file, where the docstrings are shown as "not implemented":
![image](https://user-images.githubusercontent.com/36224762/134042409-873d9a75-a09c-4654-80ce-d444f1594f5a.png)

To accomplish this, I need to be able to query eglot for the docstrings of any symbol:
```elisp
(eglot-imaginary-query-docstring "include")
```

I can't find how to do this.

Doom has a handy function for getting the documentation of the thing at point:
```elisp
  (defun night/eglot-documentation-at-point ()
    (eglot--dbind ((Hover) contents range)
        (jsonrpc-request (eglot--current-server-or-lose) :textDocument/hover
                         (eglot--TextDocumentPositionParams))
      (let ((blurb (and (not (seq-empty-p contents))
                        (eglot--hover-info contents range)))
            (hint (thing-at-point 'symbol)))
        (if blurb
            blurb
          (help-at-pt-kbd-string)))))
```

But this obviously only works for the symbol at point.

**Comment by joaotavora** [comment]
> But this obviously only works for the symbol at point.

LSP only works for the symbol at point, so I don't think you'll be easily able to make a function with that API.

[By the way, did you miss the issue creating template altogether?  People are creating issues when I ask to create conversations.  Perhaps they are simply not seeing the template where I ask this]

**Reply by NightMachinery** [reply]
> > But this obviously only works for the symbol at point.
> 
> LSP only works for the symbol at point, so I don't think you'll be easily able to make a function with that API.

I think my use case should still be very possible; Just send the input string as if it was what was at the point. The LSP server need not know the details; In fact, treating the string input as if it was typed at point is pretty much exactly what I want (so that modules and stuff are handled properly).

BTW, I think VSCode has a feature where they show a hover tooltip for symbols that are not at point. Perhaps you meant LSP works for symbols in the current file? I am not familiar with LSP, so I am a bit confused.

> [By the way, did you miss the issue creating template altogether? People are creating issues when I ask to create conversations. Perhaps they are simply not seeing the template where I ask this]

I did not use the template because it said "Bug Report," which my issue was not. You might want to add a template for "miscellaneous"?

**Reply by joaotavora** [reply]
> The LSP server need not know the details; In fact, treating the string input as if it was typed at point is pretty much exactly what I want (so that modules and stuff are handled properly).

It can't be _"as if it was typed"_ it _has_ to be typed.  You can type it and then erase it, but that probably raises other problems.

> BTW, I think VSCode has a feature where they show a hover tooltip for symbols that are not at point. Perhaps you meant LSP works for symbols in the current file? I am not familiar with LSP, so I am a bit confused.

I don't have time to explain, and perhaps I didn't understand your initial question fully.  But LSP is different, ti's not as generic an API as people sometimes take it to be.  It's made for editors (read VSCode) to work in certain ways.  So look at VSCode and study its workings for that feature, study the LSP back and forth transcript.  If it works there, it should be possible to make an Emacs program that emulates it.  Another question entirely is if that Emacs program is somehow friendly to Ivy or any other existing Emacs-land thing.

> You might want to add a template for "miscellaneous"?

Maybe. It'd be an empty template with just a link to the discussion forums.  Thanks for the idea.

**Reply by narendraj9** [reply]
I find this design choice interesting. I am curious about this because in some cases, for example, completion at point (say with the Java LSP server), we have to insert more text and then delete it later. On my machine sometimes, I can see the deletion happening. It's not a problem but feels a bit odd when text is inserted first into the buffer and then deleted. To give more context about the example here is an issue https://github.com/joaotavora/eglot/issues/668

**Reply by joaotavora** [reply]
>  we have to insert more text and then delete it later. O

I can't know 100% what you're talking about.  But that almost certainly wasn't the same thing what I was suggesting.  I'm reasonably sure the situation you're experiencing has to do with snippet insertion.

> example here is an issue #668

That issue is malformed, because the described effect can come from many places which are _not_ Eglot.  That's why I closed it. If you experience the problem and you would like to make a proper Eglot issue (following the template) where the problem can be reproduced and witnessed by me or by some developer (preferably without Java, but with Java if it has to be), then I'll be thankful to you.

**Reply by narendraj9** [reply]
Sorry for not being explicit. Maybe it is not related to this discussion. You are right, the problem I am facing seems related to snippet expansion. Hopefully, this would explain better what I mentioned earlier:

In the definition of `eglot-completion-at-point`, the completion table has the list of `label` values (https://github.com/joaotavora/eglot/blob/c17bdf6c98d6bf0f1a85f1175556e1038654402f/eglot.el#L2329). These `label` values returned from the server are usually things that are annotated (e.g. the name of the method with the type signatgure of a method). Once a completion candidate is selected, this `label` gets inserted into the buffer. After that, the `:exit-function` (defined at https://github.com/joaotavora/eglot/blob/c17bdf6c98d6bf0f1a85f1175556e1038654402f/eglot.el#L2417) deletes extra text that got inserted and also applies textEdits received from the server. For example, here is an event: 
```json
(:jsonrpc "2.0" :id 313 :method "completionItem/resolve" :params
          (:label
           #("abort(AAbort aa) : void" 0 1
             (eglot--lsp-item #1))
           :kind 2 :detail "Association.abort(AAbort aa) : void" :sortText "999999035" :insertText "abort" :insertTextFormat 2 :textEdit
           (:range
            (:start
             (:line 241 :character 11)
             :end
             (:line 241 :character 11))
            :newText "abort(${1:aa});")
           :data
           (:decl_signature "Lorg.dcm4che3.net.Association;" :signature "(Lorg.dcm4che3.net.pdu.AAbort;)V" :name "abort" :pid "0" :rid "52" :uri "file:///home/narendra/code/dcm4che/dcm4che-net/src/main/java/org/dcm4che3/net/State.java")))
```

Before sending the `completionItem/resolve` request to the LSP server, we have already inserted the `:label` text. We are eventually going to insert the expanded snippet that's present in `textEdit`.

**Reply by NightMachinery** [reply]
@joaotavora I asked around; Julia's language server exposes `julia/getDocFromWord` which accepts `{word: "symbol_here"}`. 

But unrelated to that, I just discovered that `company` already provides this functionality:
![image](https://user-images.githubusercontent.com/36224762/134241245-fe483cf3-463a-43a5-8b09-a11968cf157c.png)

Looking at the source code, I think the docstring is being extracted in the function `eglot-completion-at-point`, at `:company-doc-buffer`. It seems to me that it is indeed inserting all the candidates as text? 
```elisp
(setq cached-proxies
                      (mapcar
                       (jsonrpc-lambda
                           (&rest item &key label insertText insertTextFormat
                                  &allow-other-keys)
                         (let ((proxy
                                (cond ((and (eql insertTextFormat 2)
                                            (eglot--snippet-expansion-fn))
                                       (string-trim-left label))
                                      ((and insertText
                                            (not (string-empty-p insertText)))
                                       insertText)
                                      (t
                                       (string-trim-left label)))))
                           (unless (zerop (length proxy))
                             (put-text-property 0 1 'eglot--lsp-item item proxy))
                           proxy))
                       items))
```
------
```elisp
       :company-doc-buffer
       (lambda (proxy)
         (let* ((documentation
                 (let ((lsp-comp (get-text-property 0 'eglot--lsp-item proxy)))
                   (plist-get (funcall resolve-maybe lsp-comp) :documentation)))
                (formatted (and documentation
                                (eglot--format-markup documentation))))
           (when formatted
             (with-current-buffer (get-buffer-create " *eglot doc*")
               (erase-buffer)
               (insert formatted)
               (current-buffer)))))

```

It would be great if this API was decoupled from `company-capf`.

**Reply by joaotavora** [reply]
> It seems to me that it is indeed inserting all the candidates as text?

No it's not.  It's more complicated.  It's using `textDocument/completion` and `textDocument/resolveCompletion`. So you can get documentation for a given completion.

> It would be great if this API was decoupled from company-capf.

I don't know what that means.  And it doesn't help with your case in general.

> Julia's language server exposes julia/getDocFromWord which accepts {word: "symbol_here"}.

That works for Julia, perhaps.  Maybe you can lobby with the LSP folks that they make it a part of the protocol.

**Comment by joaotavora** [comment]
Narendra Joshi ***@***.***> writes:

> Sorry for not being explicit. Maybe it is not related to this discussion. You are right, the problem I am facing
> seems related to snippet expansion. Hopefully, this would explain
> better what I mentioned earlier:

Thanks, your analysis seems useful, but please start a new discussion
for this as it is unrelated to the current topic (unless you have
complete reproduction information to start a new issue, in that case
feel free to create one).

João
